src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -536,25 +533,26 @@ public override void List(IQueryExpression queryExpression, QueryParameters quer
 				AutoFlushIfRequired(plan.QuerySpaces);
 
 				bool success = false;
-				dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
-				try
-				{
-					plan.PerformList(queryParameters, this, results);
-					success = true;
-				}
-				catch (HibernateException)
-				{
-					// Do not call Convert on HibernateExceptions
-					throw;
-				}
-				catch (Exception e)
-				{
-					throw Convert(e, "Could not execute query");
-				}
-				finally
+				using (SuspendAutoFlush()) //stops flush being called multiple times if this method is recursively called
 				{
-					dontFlushFromFind--;
-					AfterOperation(success);
+					try
+					{
+						plan.PerformList(queryParameters, this, results);
+						success = true;
+					}
+					catch (HibernateException)
+					{
+						// Do not call Convert on HibernateExceptions
+						throw;
+					}
+					catch (Exception e)
+					{
+						throw Convert(e, "Could not execute query");
+					}
+					finally
+					{
+						AfterOperation(success);
+					}
 				}
```

Comment:
Discrepancy: the `AfterOperation(success);` was previously put after decrementing back `dontFlushFromFind`, meaning after operations could trigger auto-flush. Now it is done before leaving the new using, so before re-enabling auto-flushes.
But `AfterOperation` does in fact nothing if we are inside a transaction. And when we are not inside a transaction, auto-flush is anyway already disabled. So this change has no consequences with current code.